### PR TITLE
KATI-61_Book-Appointment-UI

### PR DESCRIPTION
**Pull request by vtnlinh95**

### CODE REVIEW

File: app/src/main/java/com/kms/cura/view/activity/BookAppointmentActivity.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    private Calendar calendar;
+    private int facilitySelected = -1;
+    private int apptLengthSelected = -1;
+    public static String DATE_SELECTED = "DATE_SELECTED";
+    public static String MONTH_SELECTED = "MONTH_SELECTED";
+    public static String YEAR_SELECTED = "YEAR_SELECTED";
+    public static String LENGTH_SELECTED = "LENGTH_SELECTED";
+    public static int REQUEST_cODE = 200;
+    public static int RESULT_OK = 400;
+    public static String HINT_TEXT = "Please choose";
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/view/activity/BookAppointmentActivity.java]: Put this in resource file

File: app/src/main/java/com/kms/cura/view/activity/BookAppointmentActivity.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    public void initTextView() {
+        txtDate = (TextView) findViewById(R.id.txtDate);
+        txtTime = (TextView) findViewById(R.id.txtTime);
+        edtComment = (EditText) findViewById(R.id.edtComment);
+    }
+
+    @Override
+    public void onClick(View v) {
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/view/activity/BookAppointmentActivity.java:119]: Split this into sub-methods: 
- chooseDate()
- chooseTime()
- back()

File: app/src/main/java/com/kms/cura/view/activity/BookAppointmentActivity.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        } else if (id == R.id.btnChoseTime) {
+            if(spnApptLength.getSelectedItemPosition()!= apptLength.size()-1 &&
+                    spnFacilities.getSelectedItemPosition() != doctorFacilityName.size() -1 && selectDate ) {
+                Intent toChooseTime = new Intent(this, ChooseTimeActivity.class);
+                Bundle bundle = createBundle();
+                toChooseTime.putExtras(bundle);
+                startActivityForResult(toChooseTime, REQUEST_cODE);
+            }
+            else {
+                Toast.makeText(this, "Facility, Date and Appointment Length can not be left empty", Toast.LENGTH_SHORT).show();
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/view/activity/BookAppointmentActivity.java]: "Facility, Date and Appointment Length can not be left empty" should be in resource file

File: app/src/main/java/com/kms/cura/view/activity/ChooseTimeActivity.java [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+        txtTimeChosen.setTag(tags[position]);
+        txtTimeChosen.setOnLongClickListener(this);
+        txtTimeChosen.setOnDragListener(this);
+        txtTimeChosen.setOnClickListener(this);
+        viewList.add(txtTimeChosen);
+        return view;
+    }
+
+    @Override
+    public boolean onDrag(View v, DragEvent event) {
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/view/activity/ChooseTimeActivity.java]: Split this method into sub-methods base on the event

File: app/src/main/java/com/kms/cura/view/activity/ChooseTimeActivity.java [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+        }
+        increasement = apptLength/2;
+        length = 24;
+        availableStatus = new int[]{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0};
+        timeFrame = getResources().getStringArray(R.array.TimeFrame24);
+    }
+
+
+    @Override
+    public void onClick(View v) {
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/view/activity/ChooseTimeActivity.java:424]: Split this methods too

File: app/src/main/java/com/kms/cura/view/activity/ChooseTimeActivity.java [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+                }
+                bundle2.putString(END_TIME, timeFrame[indexEnd + 1]);
+                back.putExtras(bundle2);
+                setResult(BookAppointmentActivity.RESULT_OK, back);
+                finish();
+            }
+        } else {
+            int position = (int) v.getTag();
+            if (selectedTime.size() == 0) {
+                colorTheSelected(position);
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/view/activity/ChooseTimeActivity.java]: return after this so we don't need else after

File: app/src/main/java/com/kms/cura/view/activity/ChooseTimeActivity.java [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+            } else {
+                selectRight = false;
+                if (!selectedTime.contains(position)) {
+                    if (checkAvailable(position))
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/view/activity/ChooseTimeActivity.java]: Combine this with upper if

File: app/src/main/java/com/kms/cura/view/activity/ChooseTimeActivity.java [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+                        return;
+                }
+                for (int i = 0; i < selectedTime.size(); ++i) {
+                    int pos = selectedTime.get(i);
+                    if (availableStatus[pos] == 0) {
+                        viewList.get(pos).setBackgroundColor(grey);
+                    } else {
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/view/activity/ChooseTimeActivity.java]: Use else if here to reduce nested if

File: app/src/main/java/com/kms/cura/view/activity/ChooseTimeActivity.java [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+        for (int i = 0; i < increasement; ++i) {
+            if (availableStatus[postion + i] == 0) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void colorTheSelected(int position) {
+        if (checkAvailable(position)) {
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/view/activity/ChooseTimeActivity.java]: Should return after checkAvailable(position) is false to reduce code inside if

**Comment by vtnlinh95** [comment]
@hieumai fixed

File: app/src/main/java/com/kms/cura/view/activity/BookAppointmentActivity.java [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+
+    }
+
+    private void chooseDate(){
+        setCurrentDate();
+        Dialog dateDialog = new DatePickerDialog(this, AlertDialog.THEME_HOLO_LIGHT, myDateListener, year, month - 1, day);
+        dateDialog.show();
+    }
+
+    private void chooseTime(){
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/view/activity/BookAppointmentActivity.java]: re-format this file

File: app/src/main/java/com/kms/cura/view/activity/BookAppointmentActivity.java [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+    private void setCurrentDate() {
+        calendar = Calendar.getInstance();
+        day = calendar.get(Calendar.DAY_OF_MONTH);
+        month = calendar.get(Calendar.MONTH) + 1;
+        year = calendar.get(Calendar.YEAR);
+
+    }
+
+    private void setSpecificDate(int day, int month, int year) {
+        if(day < this.day || month < this.month || year < this.year){
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/view/activity/BookAppointmentActivity.java]: Is `day < this.day || month < this.month || year < this.year` the correct way to determine if a day is in the past ?

Comment (hieumai) [inline app/src/main/java/com/kms/cura/view/activity/BookAppointmentActivity.java]: Also, should extract it to method like isDateInThePast() for clearer meaning

Comment (vtnlinh95) [inline app/src/main/java/com/kms/cura/view/activity/BookAppointmentActivity.java]: @hieumai  at that time this.day and this.month and this.year illustrate the current day

Comment (hieumai) [inline app/src/main/java/com/kms/cura/view/activity/BookAppointmentActivity.java]: So for example, if today is ( 30, 1, 2016 ) and the input date is ( 1, 2, 2016 ), what will be the output ?

**Comment by vtnlinh95** [comment]
@hieumai fixed

File: app/src/main/java/com/kms/cura/view/activity/BookAppointmentActivity.java [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+        if (requestCode == REQUEST_cODE) {
+            if (resultCode == RESULT_OK) {
+                Bundle time = data.getExtras();
+                txtTime.setText(time.getString(ChooseTimeActivity.START_TIME) + "-" + time.getString(ChooseTimeActivity.END_TIME));
+            }
+        }
+    }
+
+    private boolean isDateSelectedInPast(int day, int month, int year){
+        if(year >  this.year){
```
Comment (hieumai) [inline app/src/main/java/com/kms/cura/view/activity/BookAppointmentActivity.java]: you should just use the before or after method of Date class for date comparison without comparing it like this.

**Comment by vtnlinh95** [comment]
@hieumai  fixed

**Comment by hieumai** [comment]
OK from me. @huuduy216  and @toannguyen71195  please start verifying on this

**Comment by vtnlinh95** [comment]
@huuduy216  @toannguyen71195  because we have not done the search doctor yet so please click on "I forget my password" to navigate to the "Book Appointment Screen" and verify

**Comment by toannguyen71195** [comment]
@vtnlinh95 is fixing

**Comment by vtnlinh95** [comment]
@toannguyen71195 @huuduy216  fixed

**Comment by huuduy216** [comment]
@vtnlinh95 
- Maybe Make the slots a smaller (height). Please discuss with Jenny and An (maybe) on that, you can use Windows phone calendar as a reference. I think it is probably the best one I've ever used. http://cdn.slashgear.com/wp-content/uploads/2014/04/windows-phone-8-1-review-sg-33-600x352.jpg
  http://windows-phone-agenda.com/img/CalendarDayView.png
- Maybe a different button design for "Choose time"
- Maybe add divider to separate the greyed out slots too
- Change "Choose time" to "Pick appointment time" or "Pick a time slot" or something along the line
- Cannot change appointment time slot after initial click with another click

p/s: Great work in such little time. truly UI master

**Comment by toannguyen71195** [comment]
@huuduy216 You must long click to be able to drag

**Comment by huuduy216** [comment]
@toannguyen71195  Thanks for pointing that out, I've removed that from my previous comment

**Comment by vtnlinh95** [comment]
@huuduy216  the design of the choose time button is Jenny idea.
- If we allow change the appointment when clicking, the dragging won't be used anymore

**Comment by huuduy216** [comment]
@vtnlinh95  I think having 2 is better no?

**Comment by vtnlinh95** [comment]
@huuduy216  I think if we do that, no one knows we have dragging

**Comment by toannguyen71195** [comment]
@huuduy216 You can cancel it by one touch outside and choose again

**Pull request merged**